### PR TITLE
feat: Introduce webext run --devtools to open DevTools for the installed add-on right away.

### PR DESCRIPTION
**Pull request by ochameau**
This depends on bug [1787409](https://bugzilla.mozilla.org/show_bug.cgi?id=1787409) and aims at addressing issue #759.

### CODE REVIEW

**Review by rpl — CHANGES_REQUESTED** [review]
@ochameau thank you so much for looking into the improvements to the webextensions debugging, I just tried this patch locally combined with (most of) the changes from [Bug 1787409](https://bugzilla.mozilla.org/show_bug.cgi?id=1787409) (and the other patch in that stack) and it definitely provides a nicer developer experience.

It seems also (at least if I'm not mistaken) that on older Firefox versions where the changes from Bug 1787409 will not be available the additional `openDevTools` property in the installTemporaryAddon RDP request is just ignored, and so it looks that we will not need to do any runtime detection and fallback, which is also great.

Follows a couple of small tweaks to let the PR to run all tests in the CI job (fixing linting and flow error, and an unrelated test failure due to the change to the method signature that the sinon stub wasn't expecting).

Would you mind to also remove the package-lock.json change? that is unrelated (and pretty big because running `npm install` with an older npm version rewrite the entire lock file to lockfileVersion 1).

You should be able to avoid the automatic rewite of the lock file by running `npm ci` instead of `npm install`.

**Comment by ochameau** [comment]
> It seems also (at least if I'm not mistaken) that on older Firefox versions where the changes from Bug 1787409 will not be available the additional `openDevTools` property in the installTemporaryAddon RDP request is just ignored, and so it looks that we will not need to do any runtime detection and fallback, which is also great.

Yes I confirm, no backward compat is needed, unless we want to warn when using --devtools on an unsupported version.

Thanks for all the suggestion to fix all ci issues \o/

**Comment by ochameau** [comment]
I only started asking for review on https://bugzilla.mozilla.org/show_bug.cgi?id=1787409
So no rush in reviewing/landing this.

**Comment by ochameau** [comment]
I'm also trying to document this new command line argument [over there](https://github.com/mozilla/extension-workshop/pull/1398).

**Comment by ochameau** [comment]
I pushed a change to fix lint issues, but jobs are still failing whereas `yarn test` works locally.

**Comment by rpl** [comment]
> I pushed a change to fix lint issues, but jobs are still failing whereas `yarn test` works locally.

@ochameau sorry for not having got back to this sooner, I noticed it and I was meant to comment but didn't got back to this quickly enough (or at least not as quickly as you ;-))

Those failures are coming from the function, which are the ones that can be also run locally using [`npm run test-functional`](https://app.circleci.com/pipelines/github/mozilla/web-ext/1643/workflows/5cb26fc3-d663-4662-b515-25fbc12e7687/jobs/6654?invite=true#step-113-2), the reason for the failure (and related fix) is actually trivial:
- the `web-ext run` functional test is currently using a nodejs script that mocks Firefox and it doesn't currently expect the new option and so it fails, the reason is not immediately visible in the failure message but it is part of the failure logs emitted by the `tests/functional/fake-firefox-binary.js` script [here where it says `Fake Firefox received an unexpected message`](https://app.circleci.com/pipelines/github/mozilla/web-ext/1643/workflows/5cb26fc3-d663-4662-b515-25fbc12e7687/jobs/6654?invite=true#step-113-253)

The following one line patch should fix that remaining CI job failure:
```
diff --git a/tests/functional/fake-firefox-binary.js b/tests/functional/fake-firefox-binary.js
index 2f86902..00ffb3f 100755
--- a/tests/functional/fake-firefox-binary.js
+++ b/tests/functional/fake-firefox-binary.js
@@ -9,6 +9,7 @@ const REQUEST_INSTALL_ADDON = {
   to: 'fakeAddonsActor',
   type: 'installTemporaryAddon',
   addonPath: process.env.addonPath,
+  openDevTools: false, // Introduced in Firefox 106 (Bug 1787409 / Bug 1789245)
 };
 const REPLY_INSTALL_ADDON = {
   from: 'fakeAddonsActor',
```

**Comment by ochameau** [comment]
Thanks for the investigate. I hope jobs will be green now.

File: src/firefox/index.js
Comment (rpl) [inline src/firefox/index.js:176]: @ochameau I'm thinking that we may as well always log this couple of messages, instead of condition them on the verbose and devtools flags.

wdyt?

As a side-effect, that may also make codecov to don't complain about these logs emitted on `devtools === true` not being covered in tests (which are not a big deal on their own, given that its just two informative logs I wouldn't definitely block on that codecov coverage failure).

**Review by rpl — COMMENTED** [review]
@ochameau looks pretty great, I have just one more thoughts described in the inline review comment below (I'm basically wondering if we really need to condition those logs, if we think it is not necessary we can revert a couple of changes related to propagating verbose and avoid the codecov failure as a side effect).

File: src/firefox/index.js
Comment (ochameau) [reply src/firefox/index.js:176]: TBH, I'm not working enough with this tooling to have a strong opinion.

So I would prefer to defer to you/your team judgement.
I'm fine removing, or adding proper coverage.

Comment (rpl) [reply src/firefox/index.js:176]: ok, let's remove the `verbose` flag propagation and keep that log unconditioned, while condition the new one to the `devtools` flag.

Let's also change `WebExtension` to `WebExtensions` in the new logged message and let's add something like this to `test.firefox.js` to explicitly cover it and make codecov happy again:
```
import {
  consoleStream, // instance is imported to inspect logged messages
} from '../../../src/util/logger.js';

...

    it('logs link to debugging docs', async () => {
      const runner = createFakeFxRunner();
      consoleStream.flushCapturedLogs();
      consoleStream.startCapturing();

      const expectedMessage = 'More info about WebExtensions debugging:';
      const expectedURLToDocs =
        'https://extensionworkshop.com/documentation/develop/debugging/';
      await runFirefox({fxRunner: runner, devtools: false});
      assert.notOk(consoleStream.capturedMessages.find(
        (msg) => msg.includes(expectedMessage)
      ));
      assert.notOk(consoleStream.capturedMessages.find(
        (msg) => msg.includes(expectedURLToDocs)
      ));

      consoleStream.flushCapturedLogs();

      await runFirefox({fxRunner: runner, devtools: true});
      const foundMessage = consoleStream.capturedMessages.find(
        (msg) => msg.includes(expectedMessage)
      );
      const foundDocURL = consoleStream.capturedMessages.find(
        (msg) => msg.includes(expectedURLToDocs)
      );

      // Expect the logs to be found.
      assert.ok(foundMessage);
      assert.ok(foundDocURL);

      // Expected to be emitted as info level logs.
      assert.ok(foundMessage?.includes('[info]'));
      assert.ok(foundDocURL?.includes('[info]'));
    });
```

**Review by rpl — CHANGES_REQUESTED** [review]

**Comment by ochameau** [comment]
Thanks a lot for the test snippet!

**Comment by willdurand** [comment]
I am going to land this PR to avoid future conflicts.

**Pull request merged**
